File: android/app/src/main/java/dev/disobey/mango/ui/AttestationBadge.kt
package dev.disobey.mango.ui

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Help
import androidx.compose.material.icons.filled.Schedule
import androidx.compose.material.icons.filled.Shield
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import dev.disobey.mango.rust.AttestationStatus
import dev.disobey.mango.ui.theme.AttestExpiredBgDark
import dev.disobey.mango.ui.theme.AttestExpiredBgLight
import dev.disobey.mango.ui.theme.AttestExpiredBorderDark
import dev.disobey.mango.ui.theme.AttestExpiredBorderLight
import dev.disobey.mango.ui.theme.AttestExpiredTextDark
import dev.disobey.mango.ui.theme.AttestExpiredTextLight
import dev.disobey.mango.ui.theme.AttestFailedBgDark
import dev.disobey.mango.ui.theme.AttestFailedBgLight
import dev.disobey.mango.ui.theme.AttestFailedBorderDark
import dev.disobey.mango.ui.theme.AttestFailedBorderLight
import dev.disobey.mango.ui.theme.AttestFailedTextDark
import dev.disobey.mango.ui.theme.AttestFailedTextLight
import dev.disobey.mango.ui.theme.AttestUnverifiedBgDark
import dev.disobey.mango.ui.theme.AttestUnverifiedBgLight
import dev.disobey.mango.ui.theme.AttestUnverifiedBorderDark
import dev.disobey.mango.ui.theme.AttestUnverifiedBorderLight
import dev.disobey.mango.ui.theme.AttestUnverifiedTextDark
import dev.disobey.mango.ui.theme.AttestUnverifiedTextLight
import dev.disobey.mango.ui.theme.AttestVerifiedBgDark
import dev.disobey.mango.ui.theme.AttestVerifiedBgLight
import dev.disobey.mango.ui.theme.AttestVerifiedBorderDark
import dev.disobey.mango.ui.theme.AttestVerifiedBorderLight
import dev.disobey.mango.ui.theme.AttestVerifiedTextDark
import dev.disobey.mango.ui.theme.AttestVerifiedTextLight

/// Capsule-shaped attestation trust badge for the chat top bar.
/// Color and icon reflect the attestation status per UI-SPEC.
@Composable
fun AttestationBadge(
    status: AttestationStatus,
    modifier: Modifier = Modifier,
) {
    var showDetail by remember { mutableStateOf(false) }

    Surface(
        onClick = { showDetail = true },
        modifier = modifier,
        shape = androidx.compose.foundation.shape.RoundedCornerShape(50),
        color = badgeBackground(status),
        border = BorderStroke(1.dp, badgeStroke(status)),
        tonalElevation = 0.dp,
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 8.dp, vertical = 4.dp),
            horizontalArrangement = Arrangement.spacedBy(4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                imageVector = badgeIcon(status),
                contentDescription = null,
                tint = badgeForeground(status),
                modifier = Modifier.size(12.dp),
            )
            Text(
                text = badgeLabel(status),
                style = MaterialTheme.typography.labelSmall,
                color = badgeForeground(status),
            )
        }
    }

    if (showDetail) {
        AttestationDetailDialog(status = status, onDismiss = { showDetail = false })
    }
}

// MARK: - Detail Dialog

@Composable
private fun AttestationDetailDialog(status: AttestationStatus, onDismiss: () -> Unit) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Trust Status") },
        text = { Text(detailText(status)) },
        confirmButton = {
            Button(onClick = onDismiss) { Text("OK") }
        },
    )
}

// MARK: - Status Helpers

private fun badgeLabel(status: AttestationStatus): String = when (status) {
    is AttestationStatus.Verified -> "Verified"
    is AttestationStatus.Unverified -> "Not Verified"
    is AttestationStatus.Expired -> "Expired"
    is AttestationStatus.Failed -> "Failed"
}

private fun badgeIcon(status: AttestationStatus): ImageVector = when (status) {
    is AttestationStatus.Verified -> Icons.Default.CheckCircle
    is AttestationStatus.Unverified -> Icons.Default.Help
    is AttestationStatus.Expired -> Icons.Default.Schedule
    is AttestationStatus.Failed -> Icons.Default.Warning
}

@Composable
private fun badgeBackground(status: AttestationStatus): Color {
    val dark = isSystemInDarkTheme()
    return when (status) {
        is AttestationStatus.Verified ->   if (dark) AttestVerifiedBgDark else AttestVerifiedBgLight
        is AttestationStatus.Expired ->    if (dark) AttestExpiredBgDark else AttestExpiredBgLight
        is AttestationStatus.Unverified -> if (dark) AttestUnverifiedBgDark else AttestUnverifiedBgLight
        is AttestationStatus.Failed ->     if (dark) AttestFailedBgDark else AttestFailedBgLight
    }
}

@Composable
private fun badgeForeground(status: AttestationStatus): Color {
    val dark = isSystemInDarkTheme()
    return when (status) {
        is AttestationStatus.Verified ->   if (dark) AttestVerifiedTextDark else AttestVerifiedTextLight
        is AttestationStatus.Expired ->    if (dark) AttestExpiredTextDark else AttestExpiredTextLight
        is AttestationStatus.Unverified -> if (dark) AttestUnverifiedTextDark else AttestUnverifiedTextLight
        is AttestationStatus.Failed ->     if (dark) AttestFailedTextDark else AttestFailedTextLight
    }
}

@Composable
private fun badgeStroke(status: AttestationStatus): Color {
    val dark = isSystemInDarkTheme()
    return when (status) {
        is AttestationStatus.Verified ->   if (dark) AttestVerifiedBorderDark else AttestVerifiedBorderLight
        is AttestationStatus.Expired ->    if (dark) AttestExpiredBorderDark else AttestExpiredBorderLight
        is AttestationStatus.Unverified -> if (dark) AttestUnverifiedBorderDark else AttestUnverifiedBorderLight
        is AttestationStatus.Failed ->     if (dark) AttestFailedBorderDark else AttestFailedBorderLight
    }
}

private fun detailText(status: AttestationStatus): String = when (status) {
    is AttestationStatus.Verified ->
        "This conversation is routed to a Trusted Execution Environment. The TEE attestation report has been independently verified by this app."
    is AttestationStatus.Unverified ->
        "Attestation has not been checked for this backend yet."
    is AttestationStatus.Expired ->
        "The cached attestation result has expired and re-verification is needed."
    is AttestationStatus.Failed ->
        "Attestation verification failed. The backend could not prove it is running in a trusted environment.\n\nReason: ${status.reason}"
}

File: desktop/iced/src/widgets/attestation_badge.rs
use iced::widget::{button, column, container, text};
use iced::{Border, Color, Element, Length, Padding};

use mango_core::{AppState, AttestationStatus};

use crate::Message;

struct BadgeStyle {
    bg: Color,
    text_color: Color,
    border: Color,
}

fn status_style(status: &AttestationStatus, is_dark: bool) -> BadgeStyle {
    let colors = crate::theme::badge_colors(status, is_dark);
    BadgeStyle {
        bg: colors.bg,
        text_color: colors.text,
        border: colors.border,
    }
}

fn status_label(status: &AttestationStatus) -> &'static str {
    match status {
        AttestationStatus::Verified => "Verified",
        AttestationStatus::Unverified => "Not Verified",
        AttestationStatus::Expired => "Expired",
        AttestationStatus::Failed { .. } => "Failed",
    }
}

fn detail_text(status: &AttestationStatus) -> &'static str {
    match status {
        AttestationStatus::Verified => {
            "This conversation is routed to a Trusted Execution Environment. \
             The TEE attestation report has been independently verified by this app."
        }
        AttestationStatus::Unverified => {
            "Attestation has not been checked for this backend yet."
        }
        AttestationStatus::Expired => {
            "The attestation result has expired. Re-verification is pending."
        }
        AttestationStatus::Failed { .. } => {
            "Attestation verification failed. \
             The backend could not prove it is running in a trusted environment."
        }
    }
}

pub fn view<'a>(
    state: &'a AppState,
    show_detail: bool,
    is_dark: bool,
) -> Element<'a, Message> {
    // Find attestation status for the active backend
    let attest_status = state
        .active_backend_id
        .as_deref()
        .and_then(|id| state.attestation_statuses.iter().find(|e| e.backend_id == id))
        .map(|e| &e.status);

    let (label, style) = if let Some(status) = attest_status {
        (status_label(status), status_style(status, is_dark))
    } else {
        let fallback = crate::theme::badge_colors(&AttestationStatus::Unverified, is_dark);
        (
            "Not Verified",
            BadgeStyle {
                bg: fallback.bg,
                text_color: fallback.text,
                border: fallback.border,
            },
        )
    };

    let badge_text = text(label).size(12).color(style.text_color);

    let badge_container = container(badge_text)
        .padding(Padding::from([3u16, 8]))
        .style(move |_theme| container::Style {
            background: Some(iced::Background::Color(style.bg)),
            border: Border {
                color: style.border,
                width: 1.0,
                radius: 10.0.into(),
            },
            ..Default::default()
        });

    let badge_button = button(badge_container)
        .on_press(Message::ToggleAttestationDetail)
        .padding(0)
        .style(|_theme, _status| button::Style {
            background: None,
            ..Default::default()
        });

    if show_detail {
        let detail = if let Some(status) = attest_status {
            detail_text(status)
        } else {
            "Attestation has not been checked for this backend yet."
        };

        let detail_content = column![
            text("Trust Status").size(14),
            text(detail).size(12),
        ]
        .spacing(4);

        let detail_bg = if is_dark { Color::from_rgb(0.18, 0.18, 0.18) } else { Color::from_rgb(0.95, 0.95, 0.95) };
        let detail_border = if is_dark { Color::from_rgb(0.3, 0.3, 0.3) } else { Color::from_rgb(0.8, 0.8, 0.8) };

        let detail_box = container(detail_content)
            .padding(10)
            .width(Length::Fixed(280.0))
            .style(move |_theme| container::Style {
                background: Some(iced::Background::Color(detail_bg)),
                border: Border {
                    color: detail_border,
                    width: 1.0,
                    radius: 6.0.into(),
                },
                ..Default::default()
            });

        column![badge_button, detail_box].spacing(4).into()
    } else {
        badge_button.into()
    }
}
